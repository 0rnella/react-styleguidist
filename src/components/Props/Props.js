import React, { PropTypes } from 'react';

export default React.createClass({
	displayName: 'Props',
	propTypes: {
		props: PropTypes.object.isRequired
	},

	renderRows() {
		let rows = [];
		let { props } = this.props.props;
		for (var name in props) {
			var prop = props[name];
			rows.push(
				<li key={name}>
					{this.renderName(name, prop)}
					{this.renderType(prop)}
					{this.renderDescription(prop)}
				</li>
			);
		}
		return rows;
	},

	renderName(name, prop) {
		if (prop.required) {
			return name;
		}
		else {
			return (
				<span>[{name}=<code>{prop.defaultValue.value}</code>]</span>
			);
		}
	},

	renderType(prop) {
		// TODO: enums
		return `: ${prop.type.name}`;
	},

	renderDescription(prop) {
		if (prop.description) {
			return ` — ${prop.description}`;
		}
		else {
			return '';
		}
	},

	render() {
		return (
			<ul>
				{this.renderRows()}
			</ul>
		);
	}
});
