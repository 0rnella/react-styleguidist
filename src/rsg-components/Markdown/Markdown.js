import React, { PropTypes } from 'react';
import { compiler } from 'markdown-to-jsx';
import mapValues from 'lodash/mapValues';
import memoize from 'lodash/memoize';
import Styled from 'rsg-components/Styled';
import { styles as linkStyles } from 'rsg-components/Link';

// We’re explicitly specifying Webpack loaders here so we could skip specifying them in Webpack configuration.
// That way we could avoid clashes between our loaders and user loaders.
require('!!../../../loaders/style-loader!../../../loaders/css-loader!highlight.js/styles/tomorrow.css');

// Code blocks with server-side syntax highlight
function Code({ children, className }) {
	const isHighlighted = className && className.includes('lang-');
	if (isHighlighted) {
		return (
			<code className={className} dangerouslySetInnerHTML={{ __html: children }} />
		);
	}
	return (
		<code className={className}>{children}</code>
	);
}
Code.propTypes = {
	children: PropTypes.node,
	className: PropTypes.string,
};

// Custom CSS classes for each tag: <em> → <em className={s.em}> + custom components
const getBaseOverrides = memoize(classes => {
	const styleOverrides = mapValues(classes, value => ({
		props: {
			className: value,
		},
	}));

	return {
		...styleOverrides,
		code: {
			component: Code,
			props: {
				className: classes.code,
			},
		},
	};
});

// Inline mode: replace <p> (usual root component) with <span>
const getInlineOverrides = memoize(classes => {
	const overrides = getBaseOverrides(classes);

	return {
		...overrides,
		p: {
			component: 'span',
			props: {
				className: classes.base,
			},
		},
	};
});

const styles = ({ font, monospace, link, linkHover, border, codeBackground }) => ({
	base: {
		fontFamily: font,
		fontSize: 'inherit',
	},
	para: {
		fontFamily: font,
		fontSize: 'inherit',
		margin: [[0, 0, 15, 0]],
		lineHeight: 1.5,
	},
	a: linkStyles({ link, linkHover }).link,
	h1: {
		composes: '$para',
		fontSize: 36,
		fontWeight: 'normal',
	},
	h2: {
		composes: '$para',
		fontSize: 32,
		fontWeight: 'normal',
	},
	h3: {
		composes: '$para',
		fontSize: 26,
		fontWeight: 'normal',
	},
	h4: {
		composes: '$para',
		fontSize: 21,
		fontWeight: 'normal',
	},
	h5: {
		composes: '$para',
		fontSize: 16,
		fontWeight: 'normal',
	},
	h6: {
		composes: '$para',
		fontSize: 16,
		fontWeight: 'normal',
		fontStyle: 'italic',
	},
	p: {
		composes: '$para',
	},
	ul: {
		composes: '$para',
	},
	ol: {
		composes: '$para',
		listStyleType: 'decimal',
	},
	li: {
		composes: '$base',
		listStyleType: 'inherit',
	},
	input: {
		display: 'inline-block',
		margin: [[0, '0.35em', '0.25em', '-1.2em']],
		verticalAlign: 'middle',
	},
	blockquote: {
		composes: '$para',
		fontSize: 14,
		margin: [[15, 30]],
		padding: 0,
	},
	hr: {
		composes: '$para',
		borderWidth: [[0, 0, 1, 0]],
		borderColor: border,
		borderStyle: 'solid',
	},
	em: {
		composes: '$base',
		fontStyle: 'italic',
	},
	strong: {
		composes: '$base',
		fontWeight: 'bold',
	},
	code: {
		fontFamily: monospace,
		fontSize: 'inherit',
		color: 'inherit',
		background: 'transparent',
		whiteSpace: 'inherit',
	},
	pre: {
		backgroundColor: codeBackground,
		border: [[1, border, 'solid']],
		padding: [[12, 15]],
		fontSize: 12,
		borderRadius: 3,
		whiteSpace: 'pre',
	},
	table: {
		composes: '$para',
		borderCollapse: 'collapse',
	},
	thead: {
		composes: '$hr',
	},
	td: {
		fontFamily: font,
		padding: [[6, 15, 6, 0]],
		fontSize: 14,
	},
	th: {
		composes: '$td',
		fontWeight: 'bold',
	},
});

function Markdown({
	classes,
	text,
	inline,
}) {
	const overrides = inline ? getInlineOverrides(classes) : getBaseOverrides(classes);
	return compiler(text, { overrides });
}

Markdown.propTypes = {
	classes: PropTypes.object.isRequired,
	text: PropTypes.string.isRequired,
	inline: PropTypes.bool,
};

export default Styled(styles)(Markdown);
