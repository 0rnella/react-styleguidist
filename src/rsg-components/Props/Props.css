.root {
	composes: reset font from "../../styles/common.css";
	margin-bottom: 2rem;
	font-size: .9rem;
}

.heading {
	margin-bottom: .5rem;
	font-size: 1.3rem;
	font-weight: normal;
}

.table {
	width: 100%;
	border-collapse: collapse;
}
.tableHead {
	composes: border from "../../styles/colors.css";
	margin-bottom: 1.5rem;
	border-width: 0 0 1px 0;
}
.tableBody {
}
.cell {
	padding-right: 1rem;
	padding-top: .5em;
}
.cellHeading {
	padding-right: 1rem;
	padding-bottom: .5em;
	text-align:left;
}
.cellDesc {
	width: 99%;
	padding-left: 1rem;
}

.optional {
	composes: light from "../../styles/colors.css";
}

.list {
	display: inline;
	margin: 0;
	padding: 0;
	list-style-type: none;
}
.listItem {
	display: inline;
}
.listItem:after {
	content: ", ";
}
.listItem:last-child:after {
	content: "";
}
