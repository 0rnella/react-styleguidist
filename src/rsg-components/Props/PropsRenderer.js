import React from 'react';
import PropTypes from 'prop-types';
import Code from 'rsg-components/Code';
import Markdown from 'rsg-components/Markdown';
import Styled from 'rsg-components/Styled';
import Group from 'react-group';
import { unquote, getType, showSpaces } from './util';
import JsDocArguments from 'rsg-components/JsDoc/Arguments';
import JsDocDeprecated from 'rsg-components/JsDoc/Deprecated';
import JsDocLinks from 'rsg-components/JsDoc/Links';
import JsDocSince from 'rsg-components/JsDoc/Since';
import JsDocVersion from 'rsg-components/JsDoc/Version';

const styles = ({ space, color, fontFamily, fontSize }) => ({
	table: {
		width: '100%',
		borderCollapse: 'collapse',
	},
	tableHead: {
		borderBottom: [[1, color.border, 'solid']],
	},
	tableBody: {
	},
	row: {
	},
	cell: {
		color: color.base,
		paddingRight: space[2],
		paddingTop: space[1],
		verticalAlign: 'top',
		fontFamily: fontFamily.base,
		fontSize: fontSize.small,
	},
	cellHeading: {
		color: color.base,
		paddingRight: space[2],
		paddingBottom: space[1],
		textAlign: 'left',
		fontFamily: fontFamily.base,
		fontWeight: 'bold',
		fontSize: fontSize.small,
	},
	cellDesc: {
		color: color.base,
		width: '99%',
		paddingLeft: space[2],
	},
	required: {
		fontFamily: fontFamily.base,
		fontSize: fontSize.small,
		color: color.light,
	},
	name: {
		fontSize: fontSize.small,
		color: color.name,
	},
	deprecatedName: {
		fontSize: 13,
		color: light,
		textDecoration: 'line-through',
	},
	type: {
		fontSize: fontSize.small,
		color: color.type,
	},
	function: {
		fontFamily: fontFamily.base,
		fontSize: fontSize.small,
		color: color.light,
		borderBottom: [[1, 'dotted', color.lightest]],
	},
});

function renderType(type) {
	if (!type) {
		return 'unknown';
	}

	const { name } = type;

	switch (name) {
		case 'arrayOf':
			return `${type.value.name}[]`;
		case 'objectOf':
			return `{${renderType(type.value)}}`;
		case 'instanceOf':
			return type.value;
		default:
			return name;
	}
}

function renderEnum(prop) {
	if (!Array.isArray(getType(prop).value)) {
		return <span>{getType(prop).value}</span>;
	}

	const values = getType(prop).value.map(({ value }) => (
		<Code key={value}>{showSpaces(unquote(value))}</Code>
	));
	return (
		<span>One of: <Group separator=", " inline>{values}</Group></span>
	);
}

export function PropsRenderer({ classes, props }) {
	function renderRows(props) {
		const rows = [];
		for (const name in props) {
			const prop = props[name];
			rows.push(
				<tr key={name} className={classes.row}>
					<td className={classes.cell}>{renderPropName(name, prop)}</td>
					<td className={classes.cell}><Code className={classes.type}>{renderType(getType(prop))}</Code></td>
					<td className={classes.cell}>{renderDefault(prop)}</td>
					<td className={classes.cell + ' ' + classes.cellDesc}>{renderDescription(prop)}</td>
				</tr>
			);
		}
		return rows;
	}

	function renderPropName(name, prop) {
		if (prop.tags && 'deprecated' in prop.tags) {
			return (<Code className={classes.deprecatedName}>{name}</Code>);
		}

		return (<Code className={classes.name}>{name}</Code>);
	}

	function renderDefault(prop) {
		if (prop.required) {
			return (
				<span className={classes.required}>Required</span>
			);
		}
		else if (prop.defaultValue) {
			if (prop.type && prop.type.name === 'func') {
				return (
					<span className={classes.function} title={showSpaces(unquote(prop.defaultValue.value))}>Function</span>
				);
			}

			return (
				<Code>{showSpaces(unquote(prop.defaultValue.value))}</Code>
			);
		}
		return '';
	}

	function renderDescription(prop) {
		const { description } = prop;
		const extra = renderExtra(prop);

		return (
			<div>
				<JsDocDeprecated tags={prop.tags} />
				<JsDocVersion tags={prop.tags} />
				<JsDocSince tags={prop.tags} />
				<Group separator={<br />}>
					{description && <Markdown text={description} inline />}
					{extra}
				</Group>
				<JsDocLinks tags={prop.tags} />
				<JsDocArguments tags={prop.tags} />
			</div>
		);
	}

	function renderExtra(prop) {
		const type = getType(prop);

		if (!type) {
			return null;
		}
		switch (type.name) {
			case 'enum':
				return renderEnum(prop);
			case 'union':
				return renderUnion(prop);
			case 'shape':
				return renderShape(prop.type.value);
			case 'arrayOf':
				if (type.value.name === 'shape') {
					return renderShape(prop.type.value.value);
				}
				return null;
			case 'objectOf':
				if (type.value.name === 'shape') {
					return renderShape(prop.type.value.value);
				}
				return null;
			default:
				return null;
		}
	}

	function renderUnion(prop) {
		if (!Array.isArray(getType(prop).value)) {
			return <span>{getType(prop).value}</span>;
		}

		const values = getType(prop).value.map(value => (
			<Code key={value.name} className={classes.type}>{renderType(value)}</Code>
		));
		return (
			<span>One of type: <Group separator=", " inline>{values}</Group></span>
		);
	}

	function renderShape(props) {
		const rows = [];
		for (const name in props) {
			const prop = props[name];
			const defaultValue = renderDefault(prop);
			const description = prop.description;
			rows.push(
				<div key={name}>
					<Code className={classes.name}>{name}</Code>{': '}
					<Code className={classes.type}>{renderType(prop)}</Code>
					{defaultValue && ' — '}{defaultValue}
					{description && ' — '}
					{description && <Markdown text={description} inline />}
				</div>
			);
		}
		return rows;
	}

	return (
		<table className={classes.table}>
			<thead className={classes.tableHead}>
				<tr className={classes.row}>
					<th className={classes.cellHeading}>Name</th>
					<th className={classes.cellHeading}>Type</th>
					<th className={classes.cellHeading}>Default</th>
					<th className={classes.cellHeading + ' ' + classes.cellDesc}>Description</th>
				</tr>
			</thead>
			<tbody className={classes.tableBody}>
				{renderRows(props)}
			</tbody>
		</table>
	);
}

PropsRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	props: PropTypes.object.isRequired,
};

export default Styled(styles)(PropsRenderer);
