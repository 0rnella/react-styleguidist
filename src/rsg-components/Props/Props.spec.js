import test from 'ava';
import React from 'react';
import { parse } from 'react-docgen';
import Markdown from '../Markdown';
import Props, { Code } from './Props';

function render(propTypes, defaultProps = []) {
	let props = parse(`
		import { Component, PropTypes } from 'react';
		export default class Cmpnt extends Component {
			static propTypes = {
				${propTypes.join(',')}
			}
			static defaultProps = {
				${defaultProps.join(',')}
			}
			render() {
			}
		}
	`);
	return shallow(<Props props={props} />);
}

test('should render PropTypes.string', () => {
	const actual = render(['color: PropTypes.string']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>color</Code></td>
			<td><Code>string</Code></td>
			<td></td>
			<td><div /></td>
		</tr>
	);
});

test('should render PropTypes.string with a default value', () => {
	const actual = render(['color: PropTypes.string'], ['color: "pink"']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>color</Code></td>
			<td><Code>string</Code></td>
			<td><Code>pink</Code></td>
			<td><div /></td>
		</tr>
	);
});

test('should render PropTypes.string.isRequired', () => {
	const actual = render(['color: PropTypes.string.isRequired']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>color</Code></td>
			<td><Code>string</Code></td>
			<td><span>Required</span></td>
			<td><div /></td>
		</tr>
	);
});

test('should render PropTypes.arrayOf', () => {
	const actual = render(['colors: PropTypes.arrayOf(PropTypes.string)']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>colors</Code></td>
			<td><Code>string[]</Code></td>
			<td></td>
			<td><div /></td>
		</tr>
	);
});

test('should render PropTypes.instanceOf', () => {
	const actual = render(['num: PropTypes.instanceOf(Number)']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>num</Code></td>
			<td><Code>Number</Code></td>
			<td></td>
			<td><div /></td>
		</tr>
	);
});

test('should render PropTypes.shape', () => {
	const actual = render(['foo: PropTypes.shape({bar: PropTypes.number.isRequired, baz: PropTypes.any})']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>foo</Code></td>
			<td><Code>shape</Code></td>
			<td></td>
			<td>
				<div>
					<div>
						<Code>bar</Code>: <Code>number</Code> â€” <span>Required</span>
					</div>
					<div>
						<Code>baz</Code>: <Code>any</Code>
					</div>
				</div>
			</td>
		</tr>
	);
});

test('should render description in Markdown', () => {
	const actual = render(['/**\n * Label\n */\ncolor: PropTypes.string']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>color</Code></td>
			<td><Code>string</Code></td>
			<td></td>
			<td><div><Markdown text="Label" /></div></td>
		</tr>
	);
});

test('should render unknown proptype for a prop when a relevant proptype is not assigned', () => {
	const actual = render([], ['color: "pink"']);

	expect(actual.node, 'to contain',
		<tr>
			<td><Code>color</Code></td>
			<td><Code>unknown</Code></td>
			<td><Code>pink</Code></td>
			<td><div /></td>
		</tr>
	);
});
