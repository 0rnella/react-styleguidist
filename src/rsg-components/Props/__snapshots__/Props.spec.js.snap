// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render PropTypes.arrayOf 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="colors"
        />
      </td>
      <td>
        <Styled(Code)>
          string[]
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.arrayOf(PropTypes.shape) 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="foos"
        />
      </td>
      <td>
        <Styled(Code)>
          shape[]
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <div>
            <div>
              <Styled(Code)>
                bar
              </Styled(Code)>
              : 
              <Styled(Code)>
                number
              </Styled(Code)>
            </div>
            <div>
              <Styled(Code)>
                baz
              </Styled(Code)>
              : 
              <Styled(Code)>
                any
              </Styled(Code)>
            </div>
          </div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.instanceOf 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="num"
        />
      </td>
      <td>
        <Styled(Code)>
          Number
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.objectOf 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="colors"
        />
      </td>
      <td>
        <Styled(Code)>
          {string}
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.objectOf(PropTypes.shape) 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="colors"
        />
      </td>
      <td>
        <Styled(Code)>
          {shape}
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <div>
            <div>
              <Styled(Code)>
                bar
              </Styled(Code)>
              : 
              <Styled(Code)>
                number
              </Styled(Code)>
               — 
              <span>
                Required
              </span>
            </div>
            <div>
              <Styled(Code)>
                baz
              </Styled(Code)>
              : 
              <Styled(Code)>
                any
              </Styled(Code)>
            </div>
          </div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.oneOf 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="size"
        />
      </td>
      <td>
        <Styled(Code)>
          enum
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <div>
            <span>
              One of: 
              <Group
                inline={true}
                separator=", "
              >
                <Styled(Code)>
                  small
                </Styled(Code)>
                <Styled(Code)>
                  normal
                </Styled(Code)>
                <Styled(Code)>
                  large
                </Styled(Code)>
              </Group>
            </span>
          </div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.oneOfType 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="union"
        />
      </td>
      <td>
        <Styled(Code)>
          union
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <div>
            <span>
              One of type: 
              <Group
                inline={true}
                separator=", "
              >
                <Styled(Code)>
                  string
                </Styled(Code)>
                <Styled(Code)>
                  number
                </Styled(Code)>
              </Group>
            </span>
          </div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.shape 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="foo"
        />
      </td>
      <td>
        <Styled(Code)>
          shape
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <div>
            <div>
              <Styled(Code)>
                bar
              </Styled(Code)>
              : 
              <Styled(Code)>
                number
              </Styled(Code)>
               — 
              <span>
                Required
              </span>
            </div>
            <div>
              <Styled(Code)>
                baz
              </Styled(Code)>
              : 
              <Styled(Code)>
                any
              </Styled(Code)>
            </div>
          </div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.shape with description 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="foo"
        />
      </td>
      <td>
        <Styled(Code)>
          shape
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <div>
            <div>
              <Styled(Code)>
                bar
              </Styled(Code)>
              : 
              <Styled(Code)>
                number
              </Styled(Code)>
               — 
              <span>
                Required
              </span>
               — 
              <Styled(Markdown)
                inline={true}
                text="Number"
              />
            </div>
            <div>
              <Styled(Code)>
                baz
              </Styled(Code)>
              : 
              <Styled(Code)>
                any
              </Styled(Code)>
               — 
              <Styled(Markdown)
                inline={true}
                text="Any"
              />
            </div>
          </div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.string 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="color"
        />
      </td>
      <td>
        <Styled(Code)>
          string
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.string with a default value 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="color"
        />
      </td>
      <td>
        <Styled(Code)>
          string
        </Styled(Code)>
      </td>
      <td>
        <Styled(Code)>
          pink
        </Styled(Code)>
      </td>
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render PropTypes.string.isRequired 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="color"
        />
      </td>
      <td>
        <Styled(Code)>
          string
        </Styled(Code)>
      </td>
      <td>
        <span>
          Required
        </span>
      </td>
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render arguments from JsDoc tags 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="size"
        />
      </td>
      <td>
        <Styled(Code)>
          number
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <Styled(Markdown)
            text="Test description"
          />
          <JsDoc
            arg={
              Array [
                Object {
                  "description": "Converts foo to bar",
                  "name": "Foo",
                  "type": Object {
                    "name": "Array",
                  },
                },
              ]
            }
            param={
              Array [
                Object {
                  "name": "Bar",
                },
              ]
            }
          />
          <div>
            <h4>
              Arguments
            </h4>
            <Styled(Arguments)
              args={
                Array [
                  Object {
                    "description": "Converts foo to bar",
                    "name": "Foo",
                    "type": Object {
                      "name": "Array",
                    },
                  },
                  Object {
                    "name": "Bar",
                  },
                ]
              }
            />
          </div>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render description in Markdown 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="color"
        />
      </td>
      <td>
        <Styled(Code)>
          string
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <Styled(Markdown)
            text="Label"
          />
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render function body in tooltip 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="fn"
        />
      </td>
      <td>
        <Styled(Code)>
          func
        </Styled(Code)>
      </td>
      <td>
        <span
          title="(e) => console.log(e)"
        >
          Function
        </span>
      </td>
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render name as deprecated when tag deprecated is present 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={true}
          name="size"
        />
      </td>
      <td>
        <Styled(Code)>
          number
        </Styled(Code)>
      </td>
      <td />
      <td
        className="undefined undefined"
      >
        <div>
          <Styled(Markdown)
            text="Test description"
          />
          <JsDoc
            deprecated={
              Array [
                Object {
                  "description": "Do not use.",
                  "title": "deprecated",
                },
              ]
            }
          />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should render unknown proptype for a prop when a relevant proptype is not assigned 1`] = `
<table>
  <thead>
    <tr>
      <th>
        Name
      </th>
      <th>
        Type
      </th>
      <th>
        Default
      </th>
      <th
        className="undefined undefined"
      >
        Description
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <Styled(Name)
          deprecated={false}
          name="color"
        />
      </td>
      <td>
        <Styled(Code)>
          unknown
        </Styled(Code)>
      </td>
      <td>
        <Styled(Code)>
          pink
        </Styled(Code)>
      </td>
      <td
        className="undefined undefined"
      >
        <div>
          <JsDoc />
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;
